tools/fdleak.py

```diff
@@ -0,0 +1,405 @@
+}
+
+static inline void clear_pid_exit_flag(){
+  int pid = bpf_get_current_pid_tgid()>>32;
+  u8 is_exit = 0;
+  pid_exit_hash.update(&pid, &is_exit);
+}
+
+static inline void set_pid_exit_flag(){
+  int pid = bpf_get_current_pid_tgid()>>32;
+  u8 is_exit = 1;
+  pid_exit_hash.update(&pid, &is_exit);
+}
+
+static inline int fd_alloc_return(struct pt_regs *ctx) {
+  int ret_fd = PT_REGS_RC(ctx);
+  struct alloc_info_t info = {0};
+  u64 pidfd = bpf_get_current_pid_tgid() & 0xffffffff00000000L;
+  u32 pid = bpf_get_current_pid_tgid()>>32;
+  u32 tid = (u32)(bpf_get_current_pid_tgid() & 0xffffffffL);
+  struct combined_alloc_info_t *cinfo;
+  //other errors are not the monitor focus
+  if(ret_fd < 0 && ret_fd != -EMFILE)
+    return 0;
+
+  pidfd += (u32)ret_fd;
+  info.timestamp_ns = bpf_ktime_get_ns();
+  info.stack_id =
+      user_stack.get_stackid(ctx, BPF_F_USER_STACK | BPF_F_REUSE_STACKID);
+  info.tid = tid;
+  info.fd = ret_fd;
+  info.pid = pid;
+
+  clear_pid_exit_flag();
+  pidfd_allocs_hash.update(&pidfd, &info);
+  update_statistics_add(info.stack_id);
+  ##PRINT_DEBUG##
+  bpf_trace_printk("[%u] alloc fd[%d] entered \\n",info.tid, ret_fd);
+  return 0;
+}
+
+static inline void update_statistics_del(int stack_id,
+                                         u64 survivor_time_ns) {
+  struct combined_alloc_info_t *existing_cinfo;
+  struct combined_alloc_info_t cinfo = {0};
+  existing_cinfo = combined_allocs_hash.lookup(&stack_id);
+  if (existing_cinfo == NULL) {
+    bpf_trace_printk("[ERROR] missed statistics add %d\\n", stack_id);
+    return;
+  }
+  cinfo = *existing_cinfo;
+  if (cinfo.number_of_allocs > 0)
+    cinfo.number_of_allocs--;
+  if (survivor_time_ns > cinfo.last_longest_survivor) {
+    cinfo.last_longest_survivor = survivor_time_ns;
+  }
+  combined_allocs_hash.update(&stack_id, &cinfo);
+}
+
+static inline int fd_close_enter(unsigned fd) {
+  u64 pidfd = bpf_get_current_pid_tgid() & 0xffffffff00000000L;
+  struct alloc_info_t *info;
+  u64 survivor_time_ns = bpf_ktime_get_ns();
+  pidfd += (u32)fd;
+  info = pidfd_allocs_hash.lookup(&pidfd);
+  if (info == 0) {
+    return 0;
+  }
+  survivor_time_ns -= info->timestamp_ns;
+  ##PRINT_DEBUG##
+  bpf_trace_printk("[%u] free fd[%d] entered \\n",info->tid, fd);
+  pidfd_allocs_hash.delete(&pidfd);
+  update_statistics_del(info->stack_id, survivor_time_ns);
+
+  return 0;
+}
+
+// close fd
+// https://github.com/torvalds/linux/commit/8760c90
+// file: Rename __close_fd to close_fd and remove the files parameter
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)
+int kprobe____close_fd(struct pt_regs *ctx, struct files_struct *files,
+                       unsigned fd)
+#else
+int kprobe__close_fd(struct pt_regs *ctx, struct files_struct *files,
+                     unsigned int fd)
+#endif
+{
+  ##FILTER_PID##
+  FD_MONITOR_TRIGGER(fd)
+  return fd_close_enter(fd);
+}
+
+// alloc_fd
+int kretprobe__get_unused_fd_flags(struct pt_regs *ctx)
```

Comment:
You are right, `f_dupfd` does not call `get_unused_fd_flags`, so the fdleak misses `f_dupfd` in open file trace.. unfortunately, `close_fd` can be monitored close.
The reason why i don't monitor __alloc_fd is [`__alloc_fd` will merged to *static* alloc_fd](https://github.com/torvalds/linux/commit/aa384d1) . it will not be extened by fstable.  
It seems need to add a  trace of `f_dupfd`, is that right?

Reply:
Okay, I will add this. Thanks for your refine.